onnxruntime/core/platform/windows/env.cc

```diff
@@ -124,52 +147,171 @@ class WindowsEnv : public Env {
+          processor_mask |= BitOne << i;
+          processor_count += 1;
+        }
+      }
+      // when mask is set, do not visit next group
+      if (processor_mask) {
+        break;
       }
+      group_id++;
     }
-    if (!processorCoreCount)
-      ORT_THROW("Fatal error: 0 count processors from GetLogicalProcessorInformation");
-    return processorCoreCount;
-  }
-
-  std::vector<size_t> GetThreadAffinityMasks() const override {
-    auto generate_vector_of_n = [](int n) {
-      std::vector<size_t> ret(n);
-      std::iota(ret.begin(), ret.end(), 0);
-      return ret;
-    };
-    // Indeed 64 should be enough. However, it's harmless to have a little more.
-    SYSTEM_LOGICAL_PROCESSOR_INFORMATION buffer[256];
-    DWORD returnLength = sizeof(buffer);
-    if (GetLogicalProcessorInformation(buffer, &returnLength) == FALSE) {
-      return generate_vector_of_n(std::thread::hardware_concurrency());
-    }
-    std::vector<size_t> ret;
-    int count = (int)(returnLength / sizeof(SYSTEM_LOGICAL_PROCESSOR_INFORMATION));
-    for (int i = 0; i != count; ++i) {
-      if (buffer[i].Relationship == RelationProcessorCore) {
-        ret.push_back(buffer[i].ProcessorMask);
+    if (processor_count < processor_to - procssor_from + 1) {
+      LOGS_DEFAULT(ERROR) << "Processor <from> or <to> cross group boundary";
+      return {-1,0};
+    }
+    return {group_id, processor_mask};
+  }
+
+  std::pair<KAFFINITY, KAFFINITY> GetGroupAffinity(const std::vector<std::string>& processor_id_strs) {
+    if (processor_id_strs.empty()) {
+      return {-1, 0};
+    }
+    // use a set, in case of dups
+    std::set<int> processor_ids;
+    std::for_each(processor_id_strs.begin(),
+                  processor_id_strs.end(), [&](const std::string& processor_id_str) {
+                    // stoi exception will be caught in the caller
+                    if (IsDigit(processor_id_str)) {
+                      processor_ids.insert(std::stoi(processor_id_str.c_str()));
+                    } else {
+                      LOGS_DEFAULT(ERROR) << "Found non-digit processor id str: " << processor_id_str;
+                    }
+                  });
+    if (processor_ids.size() != processor_id_strs.size()) {
+      return {-1, 0};
+    }
+    int processor_id = 1;
+    int processor_count = 0;
+    KAFFINITY group_id = 0;
+    KAFFINITY processor_mask = 0;
+    for (const auto& group_info : this->group_info_vec_) {
+      for (int i = 0; i < group_info.ActiveProcessorCount; ++i, ++processor_id) {
+        if (processor_ids.count(processor_id)) {
+          processor_mask |= BitOne << i;
+          processor_count += 1;
+        }
       }
+      // when mask is set, do not visit next group
+      if (processor_mask) {
+        break;
+      }
+      group_id++;
     }
-    if (ret.empty())
-      return generate_vector_of_n(std::thread::hardware_concurrency());
-    return ret;
+    if (processor_count < processor_ids.size()) {
```

Comment:
> if (processor_count < processor_ids.size()) {

But you aren't resetting `processor_count` across groups.

Reply:
We shouldn't do that - once we found a matching group and start to set processor masks, the loop will break.
Why? Because we cannot set a affinity across group boundary.